source/common/network/dns_impl.cc

```diff
@@ -44,7 +45,11 @@ DnsResolverImpl::DnsResolverImpl(
         throw EnvoyException(
             fmt::format("DNS resolver '{}' is not an IP address", resolver->asString()));
       }
-      resolver_addrs.push_back(resolver->asString());
+      // Note that the ip()->port() may be zero if the port is not fully specified by the
+      // Address::Instance.
+      resolver_addrs.push_back(fmt::format(resolver->ip()->ipv6() ? "[{}]:{}" : "{}:{}",
+                                           resolver->ip()->addressAsString(),
+                                           resolver->ip()->port()));
```

Comment:
So I have no objection to this as-is, but I am concerned someone might come along and revert it.  I think it'd be safer to have a unit test with an Address::Instance subclass which overrides asString() and document in the test the reason we don't want to use asString() in this case.  If that turns out to be too hairy, you could instead comment explicitly that asString() is not advised and can either explain why inline or link to this PR.